marshal/claim.go

```diff
@@ -360,16 +367,36 @@ func (c *claim) messagePump() {
 		// needs to be serialized with Release, otherwise a race-condition can potentially
 		// lead to a write to a closed-channel. That's why we're using this lock. We're not
 		// using the main lock to avoid deadlocks since the write to the channel is blocking
-		// until someone consumes the message blocking all Commit operations
-		c.messagesLock.Lock()
-		if !c.Terminated() {
-			// This allocates a new Message to put the proto.Message in.
-			// TODO: This is really annoying and probably stupidly inefficient, is there any
-			// way to do this better?
-			tmp := Message(*msg)
-			c.messages <- &tmp
+		// until someone consumes the message blocking all Commit operations.
+		//
+		// This must not block -- if we hold the messagesLock for too long we will cause
+		// possible deadlocks.
+		//
+		// This allocates a new Message to put the proto.Message in.
+		// TODO: This is really annoying and probably stupidly inefficient, is there any
+		// way to do this better?
+		tmp := Message(*msg)
+		messagePending := true
+		for messagePending && !c.Terminated() {
+			messagePending = func() bool {
```

Comment:
I think it might be clearer if you did something like this

```
duration := messageBackoff.Duration()
...
for {
select {
case c.messages <- &tmp:
  break 
case <- time.After(duration):
  duration = messageBackoff.Duration()
}
messageBackoff.Reset()
``` 

This way, there won't be a time.Sleep and a time.After involved, and as soon as the channel becomes unblocked the goroutine will be able to proceed. 

However, since there's a big buffer on the channel this is mostly a matter of preference.